tests/performance/instrumentation_step_3.cc

```diff
@@ -0,0 +1,267 @@
+                            1. *                                // f(x_q)
+                            fe_values.JxW(q_index));            // dx
+        }
+      cell->get_dof_indices(local_dof_indices);
+
+      for (const unsigned int i : fe_values.dof_indices())
+        for (const unsigned int j : fe_values.dof_indices())
+          system_matrix.add(local_dof_indices[i],
+                            local_dof_indices[j],
+                            cell_matrix(i, j));
+
+      for (const unsigned int i : fe_values.dof_indices())
+        system_rhs(local_dof_indices[i]) += cell_rhs(i);
+    }
+
+
+  std::map<types::global_dof_index, double> boundary_values;
+  VectorTools::interpolate_boundary_values(dof_handler,
+                                           0,
+                                           Functions::ZeroFunction<2>(),
+                                           boundary_values);
+  MatrixTools::apply_boundary_values(boundary_values,
+                                     system_matrix,
+                                     solution,
+                                     system_rhs);
+}
+
+
+void
+Step3::solve()
+{
+  SolverControl            solver_control(1000, 1e-6 * system_rhs.l2_norm());
+  SolverCG<Vector<double>> solver(solver_control);
+  PreconditionSSOR<SparseMatrix<double>> preconditioner;
+  preconditioner.initialize(system_matrix, 1.2);
+  solver.solve(system_matrix, solution, system_rhs, preconditioner);
+}
+
+
+void
+Step3::output_results() const
+{
+  DataOut<2> data_out;
+  data_out.attach_dof_handler(dof_handler);
+  data_out.add_data_vector(solution, "solution");
+  data_out.build_patches();
+}
+
+
+Measurement
+Step3::run()
+{
+  std::map<std::string, std::uint64_t> cycle_count;
+
+  CallgrindWrapper::start_instrumentation();
+  make_grid();
+  cycle_count["make_grid"] = CallgrindWrapper::stop_instrumentation();
+
+  CallgrindWrapper::start_instrumentation();
+  setup_system();
+  cycle_count["setup_system"] = CallgrindWrapper::stop_instrumentation();
+
+  CallgrindWrapper::start_instrumentation();
+  assemble_system();
+  cycle_count["assemble_system"] = CallgrindWrapper::stop_instrumentation();
+
+  CallgrindWrapper::start_instrumentation();
+  solve();
+  cycle_count["solve"] = CallgrindWrapper::stop_instrumentation();
+
+  CallgrindWrapper::start_instrumentation();
+  output_results();
+  cycle_count["output_results"] = CallgrindWrapper::stop_instrumentation();
+
+  return {cycle_count["make_grid"],
+          cycle_count["setup_system"],
+          cycle_count["assemble_system"],
+          cycle_count["solve"],
+          cycle_count["output_results"]};
+}
+
+
+std::tuple<Metric, unsigned int, std::vector<std::string>>
+describe_measurements()
+{
+  return {Metric::instruction_count,
+          1,
+          {"make_grid",
+           "setup_system",
+           "assemble_system",
+           "solve",
+           "output_results"}};
```

Comment:
I come late to the patch (it has already been merged), but my comment about the maps was this:

The way you have implemented it, you need to keep two places in the code in sync: The place where you return a vector of cycle counts, and the place where you describe what the elements of that vector mean. This is awkward and prone to mistakes. It would be easier if the `run()` function returned self-contained information, namely a `std::map<string,CycleCountType>`. 

I have to admit that I don't like the current design. It also doesn't lend itself to hierarchical composition of cycle counting. If you let the `run()` function return a `std::map`, then it could internally call other functions that themselves create these maps, and then just merge these maps. You cannot easily do this with the current design.